src/hotspot/share/runtime/mutexLocker.cpp

```diff
@@ -201,183 +195,187 @@ static void add_mutex(Mutex* var) {
   _mutex_array[_num_mutex++] = var;
 }
 
-#define def(var, type, pri, ...) {            \
-  var = new type(Mutex::pri, #var, ##__VA_ARGS__); \
-  add_mutex(var);                             \
+#define MUTEX_STORAGE_NAME(name) name##_storage
+#define MUTEX_STORAGE(name, type) alignas(type) static uint8_t MUTEX_STORAGE_NAME(name)[sizeof(type)]
+#define MUTEX_DEF(name, type, pri, ...) {                                                       \
+  assert(name == nullptr, "Mutex/Monitor initialized twice");                                   \
+  MUTEX_STORAGE(name, type);                                                                    \
+  name = ::new(static_cast<void*>(MUTEX_STORAGE_NAME(name))) type((pri), #name, ##__VA_ARGS__); \
+  add_mutex(name);                                                                              \
 }
+#define MUTEX_DEFN(name, type, pri, ...) MUTEX_DEF(name, type, Mutex::pri, ##__VA_ARGS__)
 
 // Specify relative ranked lock
 #ifdef ASSERT
-#define defl(var, type, held_lock, ...) {         \
-  var = new type(held_lock->rank()-1, #var, ##__VA_ARGS__); \
-  add_mutex(var);                                      \
-}
+#define MUTEX_DEFL(name, type, held_lock, ...) MUTEX_DEF(name, type, (held_lock)->rank()-1, ##__VA_ARGS__)
```

Comment:
Pre-existing but what does the 'L' suffix mean here? We now have _DEF _DEFN and _DEFL and I'm not clear what the N and L signify. ??

Nit while you are here `rank()-1` -> `rank() - 1`

Reply:
Fixed the spacing to be `rank() - 1`.

As far as `N` and `L` I don't think its anything special. I chose `_DEF` to be the shared macro implementation and `N` as a short for `DEFiNition`. `L` I assume means lock, as in define using another lock's rank. But that is a guess.

Reply:
Yes, the 'l" in defl was for "lock" originally.